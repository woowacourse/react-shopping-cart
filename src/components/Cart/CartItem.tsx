import { useState } from 'react';
import {
  ProductPrice,
  ProductTitle,
  StepperContainer,
  StepperButton,
  StepperQuantity,
  ModifyRow,
  DeleteButton,
  CartItemContainer,
  CartProductImage,
  ProductRow,
  CartContent,
  CheckboxContainer,
  HiddenCheckbox,
  StyledCheckbox,
} from './Cart.styles';

function CartItem() {
  const [isChecked, setIsChecked] = useState(true);

  return (
    <CartItemContainer>
      <ModifyRow>
        <CheckboxContainer>
          <HiddenCheckbox
            type="checkbox"
            checked={isChecked}
            onChange={(e) => setIsChecked(e.target.checked)}
          />
          <StyledCheckbox checked={isChecked} />
        </CheckboxContainer>
        <DeleteButton>삭제</DeleteButton>
      </ModifyRow>

      <ProductRow>
        <CartProductImage src={''} />
        <CartContent>
          <ProductTitle>상품이름A</ProductTitle>
          <ProductPrice>35,000원</ProductPrice>
          <StepperContainer>
            <StepperButton disabled={true}>−</StepperButton>
            <StepperQuantity>2</StepperQuantity>
            <StepperButton>＋</StepperButton>
          </StepperContainer>
        </CartContent>
      </ProductRow>
    </CartItemContainer>
  );
}

export default CartItem;
