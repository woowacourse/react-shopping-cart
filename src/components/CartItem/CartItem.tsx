import { CartItemType } from "../../types/response";
import CheckBox from "../CheckBox/CheckBox";
import QuantityControlButton from "../QuantityControlButton/QuantityControlButton";
import {
  CountContainer,
  DeleteButton,
  ItemContainer,
  ItemContent,
  ItemController,
  ItemDetail,
  ItemInfo,
  ItemPrice,
  ItemTitle,
  ProductImage,
} from "./CartItem.styles";

interface CartItemProps {
  cartItem: CartItemType;
  isSelected?: boolean;
  toggleSelect?: () => void;
  increaseQuantity?: (cartId: number) => Promise<void>;
  decreaseQuantity?: (cartId: number) => Promise<void>;
  deleteCartItem?: (cartId: number) => Promise<void>;
}

function CartItem({
  cartItem,
  isSelected,
  toggleSelect,
  increaseQuantity,
  decreaseQuantity,
  deleteCartItem,
}: CartItemProps) {
  const { id: cartId, product, quantity } = cartItem;

  return (
    <>
      <div css={ItemContainer}>
        <div css={ItemController}>
          {isSelected && toggleSelect && (
            <CheckBox
              id={String(cartId)}
              isSelected={isSelected}
              onClick={toggleSelect}
            />
          )}
          {deleteCartItem && (
            <button css={DeleteButton} onClick={() => deleteCartItem(cartId)}>
              삭제
            </button>
          )}
        </div>
        <div css={ItemInfo}>
          <img
            css={ProductImage}
            src={product.imageUrl}
            onError={(error) => {
              error.currentTarget.src = "default-cartItem.png";
            }}
          ></img>
          <div css={ItemContent}>
            <div css={ItemDetail}>
              <h3 css={ItemTitle}>{product.name}</h3>
              <p css={ItemPrice}>{product.price.toLocaleString()}원</p>
            </div>
            <div css={CountContainer}>
              {decreaseQuantity && increaseQuantity ? (
                <QuantityControlButton
                  quantity={quantity}
                  decreaseQuantity={() => decreaseQuantity(cartId)}
                  increaseQuantity={() => increaseQuantity(cartId)}
                />
              ) : (
                <p>{quantity}</p>
              )}
            </div>
          </div>
        </div>
      </div>
    </>
  );
}

export default CartItem;
