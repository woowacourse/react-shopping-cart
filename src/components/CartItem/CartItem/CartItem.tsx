import useQuantityControl from "../../../hooks/cart/useQuantityControl";
import { CartItemType } from "../../../types/response";
import CheckBox from "../../CheckBox/CheckBox";
import QuantityControlButton from "../../QuantityControlButton/QuantityControlButton";
import {
  CountContainer,
  DeleteButton,
  ItemContainer,
  ItemContent,
  ItemController,
  ItemInfo,
} from "./CartItem.styles";
import CartItemImage from "../CartItemImage/CartItemImage";
import CartItemInfo from "../CartItemInfo/CartItemInfo";

interface CartItemProps {
  cartItem: CartItemType;
  fetchCartItem: () => void;
  isSelected: boolean;
  toggleSelect: () => void;
}

function CartItem({
  cartItem,
  fetchCartItem,
  isSelected,
  toggleSelect,
}: CartItemProps) {
  const { id: cartId, product, quantity: initialQuantity } = cartItem;
  const { decreaseQuantity, increaseQuantity, deleteCartItem, quantity } =
    useQuantityControl({
      initialQuantity,
      refetchCartItem: fetchCartItem,
    });

  return (
    <>
      <div css={ItemContainer}>
        <div css={ItemController}>
          <CheckBox
            id={String(cartId)}
            isSelected={isSelected}
            onClick={toggleSelect}
          />
          <button css={DeleteButton} onClick={() => deleteCartItem(cartId)}>
            삭제
          </button>
        </div>
        <div css={ItemInfo}>
          <CartItemImage image={product.imageUrl} />
          <div css={ItemContent}>
            <CartItemInfo name={product.name} price={product.price} />
            <div css={CountContainer}>
              <QuantityControlButton
                quantity={quantity}
                decreaseQuantity={() => decreaseQuantity(cartId)}
                increaseQuantity={() => increaseQuantity(cartId)}
              />
            </div>
          </div>
        </div>
      </div>
    </>
  );
}

export default CartItem;
