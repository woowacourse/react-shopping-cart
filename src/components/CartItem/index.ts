export * from './CartProduct';
export * from './CartProductContainer';
export * from './EmptyShoppingCart';
export * from './QuantitySelector';
export * from './SelectedCartProductContainer';
