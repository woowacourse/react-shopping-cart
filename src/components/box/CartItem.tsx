import styled from '@emotion/styled';
import { useState } from 'react';
import { Text } from '../common/Text/Text';
import { TrashCanIcon } from '../../assets';
import InputStepper from '../common/InputStepper/InputStepper';
import { CheckBox } from '../list/CartList';

const CartItem = () => {
  const [quantity, setQuantity] = useState(0);

  const handleSetQuantityOnInputStepper = (value: number) => {
    setQuantity(value);
  };

  return (
    <CartItemWrapper>
      <CheckBox type="checkbox" />
      <CartItemImage />
      <Text size="smallest" weight="light" color="#333333">
        [든든] 야채바삭 김말이 700g
      </Text>
      <CartItemControllerWrapper>
        <TrashCanIcon width={24} height={24} cursor="pointer" />
        <InputStepper
          size="big"
          quantity={quantity}
          setQuantity={handleSetQuantityOnInputStepper}
        />
        <Text size="minimum" weight="light" color="#333333">
          8,000원
        </Text>
      </CartItemControllerWrapper>
    </CartItemWrapper>
  );
};

export default CartItem;

const CartItemWrapper = styled.div`
  display: flex;
  justify-content: space-between;

  height: 203px;
  padding: 20px 10px;
  border-top: 1px solid #cccccc;
`;

const CartItemImage = styled.img`
  width: 144px;
  height: 147px;
`;

const CartItemControllerWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-end;

  width: 144px;
  height: 147px;
`;
