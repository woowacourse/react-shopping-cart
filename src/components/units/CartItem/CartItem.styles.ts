import styled from '@emotion/styled';

const Root = styled.li`
  display: flex;
  align-items: flex-start;
  width: 100%;
  padding: 30px 0;
  border-bottom: 1.5px solid #cccccc;
`;

const Image = styled.img`
  width: 144px;
  height: 144px;
  object-fit: contain;
  margin-left: 15px;
  box-sizing: border-box;
  border-radius: 2px;
`;

const Title = styled.div`
  flex: 1;
  margin-left: 20px;
  font-size: 20px;
  letter-spacing: 0.5px;
`;

const Option = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin-left: 20px;
`;

const DeleteButton = styled.button`
  background: none;
  border: none;
  color: inherit;
  cursor: pointer;
  margin-bottom: 20px;
`;

const Price = styled.div`
  font-size: 16px;
`;

const QuantityInputWrapper = styled.div`
  margin-bottom: 13px;
`;

export default {
  Root,
  Image,
  Title,
  Option,
  DeleteButton,
  Price,
  QuantityInputWrapper,
};
