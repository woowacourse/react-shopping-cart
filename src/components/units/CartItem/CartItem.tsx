import React from 'react';
import { useDispatch } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import { ReactComponent as DeleteIcon } from 'assets/images/delete.svg';
import noImageURL from 'assets/images/no_image.jpg';
import * as T from 'types';
import { updateQuantityRequest } from 'modules/cartItems/actions';
import CART_ITEM_QUANTITY from 'constants/cart';
import { RootState } from 'modules';
import QuantityInput from '../../shared/QuantityInput/QuantityInput';
import Checkbox from '../../shared/Checkbox/Checkbox';
import Styled from './CartItem.styles';

type CartItemProps = {
  cartItem: T.CartItem;
  onCheck: (id: number, isChecked: boolean) => void;
  onDelete: (id: T.CartItem['id']) => void;
};

const CartItem = (props: CartItemProps) => {
  const { cartItem, onCheck, onDelete } = props;
  const { id, product, quantity, checked } = cartItem;
  const { name, image, price } = product;

  const dispatch = useDispatch<ThunkDispatch<RootState, null, Action>>();

  const handleChangeInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newQuantity = event.target.valueAsNumber;

    if (newQuantity <= CART_ITEM_QUANTITY.MIN || Number.isNaN(newQuantity)) {
      dispatch(updateQuantityRequest(id, CART_ITEM_QUANTITY.MIN));
      return;
    }

    if (newQuantity >= CART_ITEM_QUANTITY.MAX) {
      dispatch(updateQuantityRequest(id, CART_ITEM_QUANTITY.MAX));
      return;
    }

    dispatch(updateQuantityRequest(id, newQuantity));
  };

  const handleIncrement = () => {
    if (quantity >= CART_ITEM_QUANTITY.MAX) return;

    dispatch(updateQuantityRequest(id, quantity + 1));
  };

  const handleDecrement = () => {
    if (quantity <= CART_ITEM_QUANTITY.MIN) return;

    dispatch(updateQuantityRequest(id, quantity - 1));
  };

  const totalPrice = price * quantity;

  const handleCheck = (event: React.ChangeEvent<HTMLInputElement>) => {
    const isChecked = event.target.checked;
    onCheck(id, isChecked);
  };

  const handleDelete = () => {
    onDelete(id);
  };

  return (
    <Styled.Root>
      <Checkbox checked={checked} onChange={handleCheck} />
      <Styled.Image src={image || noImageURL} alt={name} />
      <Styled.Title>{name}</Styled.Title>
      <Styled.Option>
        <Styled.Delete onClick={handleDelete}>
          <DeleteIcon />
        </Styled.Delete>
        <Styled.QuantityInputWrapper>
          <QuantityInput
            value={quantity}
            min={1}
            max={99}
            onChangeInput={handleChangeInput}
            onIncrease={handleIncrement}
            onDecrease={handleDecrement}
          />
        </Styled.QuantityInputWrapper>
        <Styled.Price>{totalPrice.toLocaleString('ko-KR')} Ïõê</Styled.Price>
      </Styled.Option>
    </Styled.Root>
  );
};

export default CartItem;
