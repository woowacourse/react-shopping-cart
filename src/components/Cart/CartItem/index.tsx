import * as S from './CartItem.styles';
import Counter from 'components/@common/Counter';
import Svg from 'components/@common/Svg';
import { useProductSelect } from 'hooks/useProductSelect';
import { useCheckedItems } from '../hooks/useCheckedItems';
import { Cart } from 'types';

interface CartItemProps {
  cartItem: Cart;
}

const CartItem = ({ cartItem }: CartItemProps) => {
  const { currentCartItem, remove, add, onDeleteItem } = useProductSelect(
    cartItem.product
  );
  const { checkItem, checkedItems } = useCheckedItems();
  const { product } = cartItem;

  const onCheckBoxChange = () => {
    checkItem(cartItem);
  };

  return (
    <S.CartItemWrapper>
      <S.CheckBox
        type="checkbox"
        onChange={onCheckBoxChange}
        checked={checkedItems.includes(cartItem)}
      />
      <S.CartItemImage src={product.imageUrl} alt={product.name} />
      <S.CartProductName>{product.name}</S.CartProductName>
      <S.CounterWrapper>
        <button onClick={onDeleteItem}>
          <Svg type="trash-can" width={24} height={24} />
        </button>
        <Counter
          count={currentCartItem?.quantity || 0}
          min={1}
          increment={add}
          decrement={remove}
        />
        <S.CartProductPrice>
          {product.price.toLocaleString('KR')}Ïõê
        </S.CartProductPrice>
      </S.CounterWrapper>
    </S.CartItemWrapper>
  );
};

export default CartItem;
