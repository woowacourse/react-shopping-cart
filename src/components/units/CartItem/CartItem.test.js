import React from 'react';
import { render } from '../../../test-utils';
import CartItem from './CartItem';
import defaultImageUrl from '../../../assets/images/default_product_item.png';

describe('CartItem 테스트', () => {
  test('matches snapshot', () => {
    const cartItem = {
      id: 1231234,
      product: {
        id: 123123,
        name: '맛있는 아쌈 밀크티',
        price: 10000,
        image: defaultImageUrl,
      },
      quantity: 3,
      checked: true,
    };

    const utils = render(<CartItem cartItem={cartItem} />);

    expect(utils.container).toMatchSnapshot();
  });
});
