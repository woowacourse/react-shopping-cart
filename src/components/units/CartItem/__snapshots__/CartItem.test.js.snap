// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem 테스트 matches snapshot 1`] = `
<div>
  <li
    class="css-4h9l9a-Root ezkszs76"
  >
    <label
      class="css-93sgxi-Label e1euqmb72"
    >
      <input
        checked=""
        class="css-1igejdc-Checkbox e1euqmb70"
        type="checkbox"
      />
      <span
        class="css-yz3jig-CheckMark e1euqmb73"
      />
      <span
        class="css-14nxd82-Text e1euqmb71"
      />
    </label>
    <img
      alt="item-image"
      class="css-q92kmb-Image ezkszs75"
      src="no_image.jpg"
    />
    <div
      class="css-dyy27e-Title ezkszs74"
    />
    <div
      class="css-1nr2lfb-Option ezkszs73"
    >
      <button
        class="css-obmvoo-DeleteButton ezkszs72"
      >
        <svg>
          delete.svg
        </svg>
      </button>
      <div
        class="css-gv3xyc-QuantityInputWrapper ezkszs70"
      >
        <div
          class="css-1sxgpgf-Root e1crnf635"
        >
          <input
            class="css-1l4nyb3-Input e1crnf634"
            max="99"
            min="1"
            type="number"
            value="3"
          />
          <div
            class="css-ajeb31-Control e1crnf633"
          >
            <button
              class="css-fv4lij-ArrowWrapper e1crnf632"
            >
              <img
                alt="arrow-up"
                class="css-le1t2-ArrowUpImage e1crnf631"
                src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQBAMAAADt3eJSAAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAASUExURf///zMzM0RERO/v73d3d7u7u2XEkIgAAAA8SURBVAjXY2DABlgDoAwjZSjDUQQqIygIkTMSFITIOQoKikBlIHJGIIYyRAYsB5YByRlBGMoQGag+VAAAlV0GYY4qHg4AAAAASUVORK5CYII="
              />
            </button>
            <button
              class="css-fv4lij-ArrowWrapper e1crnf632"
            >
              <img
                alt="arrow-down"
                class="css-ygv2m9-ArrowUpImage-ArrowDownImage e1crnf630"
                src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQBAMAAADt3eJSAAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAASUExURf///zMzM0RERO/v73d3d7u7u2XEkIgAAAA8SURBVAjXY2DABlgDoAwjZSjDUQQqIygIkTMSFITIOQoKikBlIHJGIIYyRAYsB5YByRlBGMoQGag+VAAAlV0GYY4qHg4AAAAASUVORK5CYII="
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="css-1b6jikt-Price ezkszs71"
      >
        NaN
         원
      </div>
    </div>
  </li>
</div>
`;
